functions/snippets/src/main/java/com/example/functions/OcrProcessImage.java

```diff
@@ -0,0 +1,161 @@
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+// [END functions_ocr_process]
+
+public class OcrProcessImage implements BackgroundFunction<GcsEvent> {
+  // TODO<developer> set these environment variables
+  private static final String PROJECT_ID = System.getenv("GCP_PROJECT");
+  private static final String TRANSLATE_TOPIC_NAME = System.getenv("TRANSLATE_TOPIC");
+  private static final String[] TO_LANGS = System.getenv("TO_LANG").split(",");
+
+  private static final Logger LOGGER = Logger.getLogger(OcrProcessImage.class.getName());
+  private static final String LOCATION_NAME = LocationName.of(PROJECT_ID, "global").toString();
+  private Publisher publisher;
+  private static final Gson gson = new Gson();
+
+  public OcrProcessImage() throws IOException {
+    publisher = Publisher.newBuilder(
+        ProjectTopicName.of(PROJECT_ID, TRANSLATE_TOPIC_NAME)).build();
+  }
+
+  // [START functions_ocr_process]
+  @Override
+  public void accept(GcsEvent gcsEvent, Context context) {
+
+    // Validate parameters
+    String bucket = gcsEvent.getBucket();
+    if (bucket == null) {
+      throw new IllegalArgumentException("Missing bucket parameter");
+    }
+    String filename = gcsEvent.getName();
+    if (filename == null) {
+      throw new IllegalArgumentException("Missing name parameter");
+    }
+
+    detectText(bucket, filename);
+  }
+  // [END functions_ocr_process]
+
+  // [START functions_ocr_detect]
+  private void detectText(String bucket, String filename) {
+    LOGGER.info("Looking for text in image " + filename);
+
+    List<AnnotateImageRequest> visionRequests = new ArrayList<>();
+    String gcsPath = String.format("gs://%s/%s", bucket, filename);
+
+    ImageSource imgSource = ImageSource.newBuilder().setGcsImageUri(gcsPath).build();
+    Image img = Image.newBuilder().setSource(imgSource).build();
+
+    Feature textFeature = Feature.newBuilder().setType(Feature.Type.TEXT_DETECTION).build();
+    AnnotateImageRequest visionRequest =
+        AnnotateImageRequest.newBuilder().addFeatures(textFeature).setImage(img).build();
+    visionRequests.add(visionRequest);
+
+    // Detect text in an image using the Cloud Vision API
+    AnnotateImageResponse visionResponse;
+    try (ImageAnnotatorClient client = ImageAnnotatorClient.create()) {
+      visionResponse = client.batchAnnotateImages(visionRequests).getResponses(0);
+      if (visionResponse == null || !visionResponse.hasFullTextAnnotation()) {
+        LOGGER.info(String.format("Image %s contains no text", filename));
+        return;
+      }
+
+      if (visionResponse.hasError()) {
+        // Throw IOException; to be handled by catch block below
+        throw new IOException(visionResponse.getError().getMessage());
```

Comment:
We shouldn't rewrap errors like this - the `visionResponse` isn't an `IOException`.

Reply:
Added a separate logging statement instead.